src/s3/commands/uploadFile.ts

```diff
@@ -104,104 +140,244 @@ export async function uploadFileCommand(
         },
-        progress => {
-            return s3Client.uploadFile({
-                bucketName: bucketName,
-                key: key,
-                fileLocation,
-                progressListener: progressReporter({ progress, totalBytes: fileSizeBytes }),
+        async (progress, token) => {
+            let uploadedCount: number = 0
+            let requestIdx: number = 0
+            const failedRequests: UploadRequest[] = []
+
+            token.onCancellationRequested(e => {
+                if (uploadRequests[requestIdx].ongoingUpload) {
+                    uploadRequests[requestIdx].ongoingUpload?.abort()
+                }
+                return failedRequests
             })
+
+            token.isCancellationRequested
+            const progressListener = progressReporter({ progress, totalBytes: uploadRequests.length })
+            //handle the loop here
+            //modify uploadwithProgress() to be for a single file again
+
+            while (!token.isCancellationRequested && requestIdx < uploadRequests.length) {
+                const request = uploadRequests[requestIdx]
+                const fileName = path.basename(request.key)
+                const destinationPath = readablePath({ bucket: { name: request.bucketName }, path: request.key })
+                const destinationNoFile = destinationPath.slice(0, destinationPath.lastIndexOf('/'))
+                showOutputMessage(
+                    localize('AWS.s3.uploadFile.startUpload', 'Uploading file {0} to {1}', fileName, destinationPath),
+                    outputChannel
+                )
+
+                const uploadResult = await uploadWithProgress(request, window, outputChannel)
+
+                if (uploadResult) {
+                    //this request failed to upload
+                    failedRequests.push(uploadResult)
+                } else {
+                    uploadedCount += 1
+                }
+
+                showOutputMessage(
+                    localize(
+                        'AWS.s3.uploadFile.progressReport',
+                        '{0}/{1} files uploaded to {2}',
+                        uploadedCount,
+                        uploadRequests.length,
+                        destinationNoFile
+                    ),
+                    outputChannel
+                )
+
+                requestIdx += 1
+                if (progressListener) {
+                    progressListener(requestIdx)
+                }
+            }
+
+            return failedRequests.concat(uploadRequests.slice(requestIdx))
         }
     )
+    return response
+}
+
+/**
+ * Uploads a single request to s3 with a progress window
+ *
+ * @param request File to be uploaded
+ * @returns The same request if failed, undefined otherwise
+ */
+async function uploadWithProgress(
+    request: UploadRequest,
+    window = Window.vscode(),
+    outputChannel = ext.outputChannel
+): Promise<UploadRequest | undefined> {
```

Comment:
I don't think this is returning anything?

Reply:
It is, it returns undefined if success, the same request if failed